README
FUN a.k.a Frugalware Update Notifier
====================================

FUN is a client-server application with a sole purpose of notifying the 
user of new package updates.

Note: As of version 0.3.0, FUN is now also capable of displaying notifications
when some news is available on the frugalware.org website. So, FUN is no longer
just an update notifier, but a news notifier too. Future versions will also
support notifications when a security fix is released.

fund (fun daemon) runs in the background waiting for commands from the
client. The fun client is a small application that resides in the user's 
system tray and periodically sends certain commands to the fun daemon to get
information about package updates and other package related information.

The fun daemon (fund) is written in C and uses DBUS for communication. 
The fun client (fun) is written in C using the GTK+ toolkit.

fun is not a DE specific application and it can work on GNOME/XFCE/KDE
or any other DE provided the following requirements are satisfied.

Requirements
============
* pacman-g2 >= 3.6.4
* dbus >= 1.1.2
* dbus-glib >= 0.73
* gtk+2 >= 2.12.0
* libglade-2.0 >= 2.6
* libnotify >= 0.4.4
* libcurl
* libnxml

Using FUN
=========
You need to start the daemon by running the following command as root:
# service fun start

Once the daemon is started, you may launch the client by running the
following command:
$ fun

(A shortcut to FUN is also available in your respective DEs)

Configuring FUN
===============
FUN stores it's configuration in /home/user/.funrc file. Although you
can configure the various options directly from within FUN, you may even
edit the .funrc file manually. Currently, fun supports the following
three configuration options:

1)	update_interval

	This specifies the time in minutes after which FUN should check for
	package updates.
	Example: If you want FUN to check for new updates every hour, you
	should set update_interval=60 in the configuration file.

2)	notification_timeout

	This specifies the number of seconds for which the notification
	tooltip (bubble) will be displayed once an update is available.
	The range is 1 to 30 seconds.

3)	gfpm_launcher

	FUN doesn't have any inherent capability of updating your system
	once new packages are available. Once the user is notified of new
	updates, he/she may launch Gfpm (Graphical Frugalware Package
	Manager) to download and install the new updates.
	Updating packages on the system requires Gfpm to be run with 
	administrator (root) privileges. This can be achieved in three ways - gksu/kdesu/sudo.

	If you're using GNOME/XFCE, you probably need gksu. On KDE, you'll need
	kdesu. If you have setup sudo (see man sudo) then you may use sudo.

	This setting lets FUN to run Gfpm with one of the above mentioned
	ways. Possible values for this setting are
	gksu
	kdesu
	sudo

	Example: gfpm_launcher=gksu  (If you're using GNOME/XFCE)

4)	news_enabled

	If set to true, FUN will also notify the user in case of a new News entry
	is available on the frugalware.org homepage.
	News notification will be disabled if this is set to false.
	
	Example: news_enabled=true

5)	news_interval

	This specifies the time in minutes after which FUN should check for
	news updates on frugalware.org. This setting has no effect if 
	news_enabled is set to false.
	Example: If you want FUN to check for new NEWS updates every hour, you
	should set news_interval=60 in the configuration file.

6)	news_browser

	This setting sets the default browser to be used when the user clicks
	"Visit Link" on a news item.
	Supported browsers are: Firefox, Epiphany, Opera, Konqueror
	
	Example: news_browser=Firefox
	


For Developers
==============
If someone is interested in developing a client for FUN, following are
the services currently provided by FUN daemon.

1) TestService
This command is basically used to test whether the client can connect to
the FUN daemon. Before sending any other commands, make sure you test
connectivity to the daemon using TestService.

2) PerformUpdate
This command checks for package updates. It tells the daemon to perform
a repository sync and check if any updates are available. It returns the
list of available package updates as a string or otherwise NULL.

3) GetPackageInfo
This command is used to get extra information about an available package
update, like it's version, description etc. It accepts one argument, the
package name for which information is needed. It returns two parameters
- the package's latest version and it's description as strings.

NOTE: Service names are subject to change

Bugs
====
If you encounter any bug(s) in FUN, please report them at
http://www.bugs.frugalware.org/

Author
======
Priyank M. Gosalia <priyankmg@gmail.com>




NEWS
VERSION					DESCRIPTION
---------------------------------------------------------------------------------------------------------
0.3.4 (2013-06-25) - fix build and remove linkage to fwutil
0.3.3 (2011-09-06)              - Fix compiling with libnotify and curl
0.3.2 (2009-03-12)		- BUGFIX (#3653): FUN becomes unresponsive on launching GFpm
				- BUGFIX: FUN shows no browser(s) if Firefox was not installed
				- Other minor fixes (see ChangeLog)
				- New translation: (Vietnamese) vi_VI
0.3.1 (2008-09-20)		- Dropped the old method of rss fetching using libxml+curl
				- Fun now makes use of libnxml for news notifier
				- BUGFIX (#3102): FUN crashes after a while
				- Minimize the main window to system tray on delete event
0.3.0 (2008-04-30)		- FUN now supports News notifications.
				  Greatly reduced FUN's startup time.
0.2.2 (2008-03-23)		- BUGFIX (#2883): FUN always checks for updates in the -current repo even on
				  -stable systems. NOTE: This affects only -stable users.
0.2.1 (2008-03-02)		- Translation updates
				- New translation: (Czech) cs_CZ
0.2.0 (2008-01-01)		- Migrated to libnotify for displaying notification popups.
				- Got rid of EggTrayIcon. FUN client now uses GtkStatusIcon.
				- FUN client now performs a first-run update check everytime it's started.
				  This is needed because otherwise the very first update check will mostly fail.
				- Other minor fixes.
				- New translations: (Danish) da_DK, (Spanish) es_ES
				- Updated translations: (Hungarian) hu_HU, (French) fr_FR

0.1.0 (2007-11-21)		- Initial Release



src/fun-config.c
/*
 *  fun-config.c for fun
 *
 *  Copyright (C) 2007-2009 by Priyank Gosalia <priyankmg@gmail.com>
 *  Portions of this code are borrowed from gimmix
 *  gimmix is Copyright (C) 2006-2007 Priyank Gosalia <priyankmg@gmail.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <gtk/gtk.h>
#include "fun-config.h"

#define CONFIG_FILE			".funrc"
#define UPDATE_INTERVAL_DEFAULT		"60"
#define NEWS_INTERVAL_DEFAULT		"30"
#define NOTIFICATION_TIMEOUT_DEFAULT	"5"
#define GFPM_LAUNCHER_DEFAULT		"sudo"

static ConfigFile conf;

static gchar *browserlist[] = 	{	"Firefox",	"/usr/bin/firefox",
					"Epiphany",	"/usr/bin/epiphany",
					"Opera",	"/usr/bin/opera",
					"Konqueror",	"/usr/bin/konqueror",
					NULL
				};

GList*
fun_config_get_available_browsers (void)
{
	guint	i = 0;
	guint   j = 0;
	GList	*ret = NULL;
	
	while (browserlist[i] != NULL)
	{
		j = i+1;
		if (g_file_test(browserlist[j],G_FILE_TEST_EXISTS))
			ret = g_list_append (ret, g_strdup(browserlist[i]));
		i+=2;
	}
	
	return ret;
}

char*
fun_config_get_browser_path (const char *name)
{
	char	*ret = NULL;
	guint	i = 0;

	while (browserlist[i] != NULL)
	{
		if (!strcmp(browserlist[i],name))
		{
			ret = g_strdup (browserlist[++i]);
			break;
		}
		i++;
	}

	return ret;
}

void
fun_config_init (void)
{
	char	*rcfile = NULL;
	gchar	*browser = NULL;
	GList	*browsers = NULL;

	cfg_init_config_file_struct (&conf);
	cfg_add_key (&conf, "update_interval", "60");
	cfg_add_key (&conf, "notification_timeout", "5");
	cfg_add_key (&conf, "gfpm_launcher", "sudo");
	cfg_add_key (&conf, "news_enabled", "true");
	cfg_add_key (&conf, "news_interval", "30");

	/* set the default browser */
	browsers = fun_config_get_available_browsers ();
	if (browsers != NULL)
	{
		browser = g_strdup_printf (browsers->data);
		cfg_add_key (&conf, "news_browser", browser);
		g_list_free (browsers);
	}
	rcfile = cfg_get_path_to_config_file (CONFIG_FILE);
	if (cfg_read_config_file (&conf, rcfile) != 0)
	{
		goto cleanup;
	}
	else
	{
		if (!fun_config_get_value_int("update_interval"))
			cfg_add_key (&conf, "update_interval", UPDATE_INTERVAL_DEFAULT);
		if (!fun_config_get_value_int("news_interval"))
			cfg_add_key (&conf, "news_interval", NEWS_INTERVAL_DEFAULT);
		if (!fun_config_get_value_int("notification_timeout"))
			cfg_add_key (&conf, "notification_timeout", NOTIFICATION_TIMEOUT_DEFAULT);
		if (!fun_config_get_value_string("gfpm_launcher"))
			cfg_add_key (&conf, "gfpm_launcher", GFPM_LAUNCHER_DEFAULT);
		if (!fun_config_get_value_string("news_browser") && browser!=NULL)
			cfg_add_key (&conf, "news_browser", browser);	
	}
	cleanup:
	fun_config_save ();
	g_free (rcfile);
	g_free (browser);

	return;
}

char *
fun_config_get_value_string (const char *key)
{
	char *ret = NULL;

	ret = cfg_get_key_value (conf, (char*)key);

	return ret;
}

int
fun_config_get_value_int (const char *key)
{
	int ret = -1;

	ret = atoi (cfg_get_key_value (conf, (char*)key));

	return ret;
}

void
fun_config_set_value_string (const char *key, char *value)
{
	char *val = g_strdup_printf (value);
	cfg_add_key (&conf, (char*) key, value);
	g_free (val);

	return;
}

void
fun_config_set_value_int (const char *key, int value)
{
	char *val = g_strdup_printf ("%d", value);
	cfg_add_key (&conf, (char*) key, val);
	g_free (val);
	
	return;
}

void
fun_config_save (void)
{
	char *rcfile;
	
	rcfile = cfg_get_path_to_config_file (CONFIG_FILE);
	cfg_write_config_file (&conf, rcfile);
	chmod (rcfile, S_IRUSR|S_IWUSR);
	g_free (rcfile);
	
	return;
}

bool
fun_config_exists (void)
{
	char *config_file = NULL;
	bool status;
	
	config_file = cfg_get_path_to_config_file (CONFIG_FILE);
	if (g_file_test(config_file, G_FILE_TEST_EXISTS))
		status = true;
	else
		status = false;

	free (config_file);
	return status;
}

void
fun_config_free (void)
{
	cfg_free_config_file_struct (&conf);
	
	return;
}

